WalletWasabi/CoinJoin/Coordinator/Coordinator.cs

```diff
@@ -322,6 +326,43 @@ private async void Round_StatusChangedAsync(object sender, CoordinatorRoundStatu
 								}
 							}
 						}
+
+						var newIntervalTriggered = Stats.Register(round.CoinJoin);
+						// Adjust timeout: https://github.com/zkSNACKs/WalletWasabi/issues/2940
+						if (newIntervalTriggered && Stats.GetAverageWcqs().Count() > 2)
+						{
+							var lastTwo = Stats.GetAverageWcqs().TakeLast(2).ToArray();
+							var older = lastTwo[0];
+							var newer = lastTwo[1];
+
+							// If the older is better, then double down, else flip.
+							if (older > newer)
+							{
+								if (Stats.LastTimeoutAdjustment == TimeoutAdjustment.Up)
+								{
+									RoundConfig.InputRegistrationTimeout = Math.Max(60, RoundConfig.InputRegistrationTimeout * (4 / 3));
+								}
+								else
+								{
+									RoundConfig.InputRegistrationTimeout = Math.Max(60, RoundConfig.InputRegistrationTimeout * (2 / 3));
```

Comment:
Define a min and max value constant for timeout. Do not set `RoundConfig.InputRegistrationTimeout` at many places. 
1. Create a method which will calculate the new value.
2. Limit that value with a sanity check.
3. Update `RoundConfig.InputRegistrationTimeout` with the result.

Reply:
Going down with timeout, you must also consider the mempool-chain problem and the network fee growth. Somehow... If your algorithm finds out that 5 minutes is the optimal CJ timeout period regarding WCQ we will might bump into unknown issues. Put a lower limit there too.